# Task 4.8 Grouping and Aggregating Data

In [2]:
# importing libraries
import pandas as pd
import numpy as np
import os

In [7]:
# import dataset
df = pd.read_pickle(r'C:\Users\littl\Career Foundry\06.2022 Instacart Basket Analysis\02 - Data\Prepared Data\4.7_orders_products_merged.pkl')

In [8]:
# viewing data information
df.shape

(32406041, 20)

In [9]:
df.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

   Unnamed: 0 product_name  aisle_id  department_id  prices _merge  \
0         195         Soda        77              7     9.0   both   
1         195         Soda        77              7     9.0   both   
2         195         Soda        77              7     9.0   both   
3         195         Soda        77              7     9.0   both   
4         195         Soda        77              7     9.0   both   

     price_range_loc     Busiest day  Busiest days     Most Orders  
0  Mid-range product  Regularly busy  Regular days  Average Orders  
1  Mid-range product  Regularly busy  Slowest days  Average Orders  
2  Mid-range product  Regularly busy  Slowest days     Most Orders  
3  Mid-range product      Least busy  Slowest days  Average Orders  
4  Mid-range product      Least busy  Slowest days     Most Orders

## Grouping and Aggregating Data

#### Question 2. Find the mean of the order number by department for the entire dataframe. 

In [10]:
# in order to aggregate data, start by grouping then aggragate
df.groupby('department_id').order_number.agg(['mean'])

mean
department_id           
1              15.457838
2              17.277920
3              17.170395
4              17.811403
5              15.213779
6              16.439806
7              17.225802
8              15.340650
9              15.895474
10             20.197148
11             16.171095
12             15.887671
13             16.583536
14             16.771241
15             16.165037
16             17.665606
17             15.694469
18             19.310397
19             17.177343
20             16.473447
21             22.902379

#### Question 3: Comparing results from subset of 1,000,000 rows. We anticipate that with such a large subset that the comparisons will be similar; however, we find the following: 

* Departments:  The subset only had 8 departments, whereas the entire dataset has 21 departments
* Mean: The mean is slightly higher overall for the subset than the entire dataset. 
* Department 17:  In the subset department 17 has a mean of 11.29, which is significantly lower than all the others; however, when we run the entire data set, Department 17, shoots up to a mean of 15.69. 

## Loyalty Flag

#### Question 4: Create a loyalty flag for existing customers using the transform and loc() function. 

In [11]:
# creating new column for max_orders per customer
df['max_order'] = df.groupby(['user_id'])['order_number'].transform(np.max)

In [12]:
# checking output
df.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  ...  \
0                     NaN         196                  1          0  ...   
1                    15.0         196                  1          1  ...   
2                    21.0         196                  1          1  ...   
3                    29.0         196                  1          1  ...   
4                    28.0         196                  1          1  ...   

   product_name aisle_id  department_id  prices  _merge    price_range_loc  \
0          Soda       77              7     9.0    both  Mid-range product   
1          Soda       77              7     9.0    both  Mid-range product   
2          Soda       77              7     9.0    both  Mid-range product   
3          Soda       77              7     9.0    both  Mid-range product   
4          Soda       77              7     9.0    both  Mid-range product   

      Busiest day  Busiest days     Most Orders max_order  
0  Regularly busy  Regular days  Average Orders        10  
1  Regularly busy  Slowest days  Average Orders        10  
2  Regularly busy  Slowest days     Most Orders        10  
3      Least busy  Slowest days  Average Orders        10  
4      Least busy  Slowest days     Most Orders        10  

[5 rows x 21 columns]

using the loc() function to group customers

In [13]:
# using the loc() function to group customers who have over 40 orders as "Loyal Customer"
df.loc[df['max_order'] > 40, 'loyalty_flag'] = 'Loyal Customer'

In [14]:
# using the loc() function to group customers who have less than 40, but more than 10 as 'Regular Customer'
df.loc[(df['max_order'] <= 40) & (df['max_order'] > 10), 'loyalty_flag'] = 'Regular Customer'

In [15]:
# using the loc() function to group customers who have 10 or less orders as 'New Customer'
df.loc[df['max_order'] <= 10, 'loyalty_flag'] = 'New Customer'

In [16]:
# cheking output/frequencies of flags
df['loyalty_flag'].value_counts(dropna = False)

Regular Customer    15877395
Loyal Customer      10284396
New Customer         6244250
Name: loyalty_flag, dtype: int64

In [17]:
# viewing output
df.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  ...  \
0                     NaN         196                  1          0  ...   
1                    15.0         196                  1          1  ...   
2                    21.0         196                  1          1  ...   
3                    29.0         196                  1          1  ...   
4                    28.0         196                  1          1  ...   

   aisle_id department_id  prices  _merge    price_range_loc     Busiest day  \
0        77             7     9.0    both  Mid-range product  Regularly busy   
1        77             7     9.0    both  Mid-range product  Regularly busy   
2        77             7     9.0    both  Mid-range product  Regularly busy   
3        77             7     9.0    both  Mid-range product      Least busy   
4        77             7     9.0    both  Mid-range product      Least busy   

   Busiest days     Most Orders max_order  loyalty_flag  
0  Regular days  Average Orders        10  New Customer  
1  Slowest days  Average Orders        10  New Customer  
2  Slowest days     Most Orders        10  New Customer  
3  Slowest days  Average Orders        10  New Customer  
4  Slowest days     Most Orders        10  New Customer  

[5 rows x 22 columns]

### Question 5: Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). Note: What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [18]:
# checking the average price by each customer group
df.groupby('loyalty_flag').agg({'prices': ['mean', 'max', 'min']})

prices              
                       mean      max  min
loyalty_flag                             
Loyal Customer    10.386229  99999.0  1.0
New Customer      13.294406  99999.0  1.0
Regular Customer  12.495501  99999.0  1.0

#### Findings:  On average loyal customers spend the least with average pricing of 10.38, while newest customers spend the most at 13.29.  

## Spending Flag

#### Question 6: Create a spending flag for each user based on the average price across all their orders using the following criteria:
* If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
* If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [19]:
# creating a new column for average prices for each user
df['avg_price']=df.groupby(['user_id'])['prices'].transform(np.mean)

In [20]:
df.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  ...  \
0                     NaN         196                  1          0  ...   
1                    15.0         196                  1          1  ...   
2                    21.0         196                  1          1  ...   
3                    29.0         196                  1          1  ...   
4                    28.0         196                  1          1  ...   

   department_id prices  _merge    price_range_loc     Busiest day  \
0              7    9.0    both  Mid-range product  Regularly busy   
1              7    9.0    both  Mid-range product  Regularly busy   
2              7    9.0    both  Mid-range product  Regularly busy   
3              7    9.0    both  Mid-range product      Least busy   
4              7    9.0    both  Mid-range product      Least busy   

   Busiest days     Most Orders max_order  loyalty_flag avg_price  
0  Regular days  Average Orders        10  New Customer  6.367797  
1  Slowest days  Average Orders        10  New Customer  6.367797  
2  Slowest days     Most Orders        10  New Customer  6.367797  
3  Slowest days  Average Orders        10  New Customer  6.367797  
4  Slowest days     Most Orders        10  New Customer  6.367797  

[5 rows x 23 columns]

In [21]:
# creating a flag for customers whose average order price is is great than 10 using loc()
df.loc[df['avg_price'] >= 10, 'spending_flag'] = 'High Spender'

In [22]:
# creating a flag for customers whose average order price is less than 10 using loc()
df.loc[df['avg_price'] < 10, 'spending_flag'] = 'Low Spender'

In [23]:
# checking output
df['spending_flag'].value_counts(dropna=False)

Low Spender     31771789
High Spender      634252
Name: spending_flag, dtype: int64

In [24]:
df.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  ...  \
0                     NaN         196                  1          0  ...   
1                    15.0         196                  1          1  ...   
2                    21.0         196                  1          1  ...   
3                    29.0         196                  1          1  ...   
4                    28.0         196                  1          1  ...   

   prices _merge    price_range_loc     Busiest day  Busiest days  \
0     9.0   both  Mid-range product  Regularly busy  Regular days   
1     9.0   both  Mid-range product  Regularly busy  Slowest days   
2     9.0   both  Mid-range product  Regularly busy  Slowest days   
3     9.0   both  Mid-range product      Least busy  Slowest days   
4     9.0   both  Mid-range product      Least busy  Slowest days   

      Most Orders max_order  loyalty_flag avg_price spending_flag  
0  Average Orders        10  New Customer  6.367797   Low Spender  
1  Average Orders        10  New Customer  6.367797   Low Spender  
2     Most Orders        10  New Customer  6.367797   Low Spender  
3  Average Orders        10  New Customer  6.367797   Low Spender  
4     Most Orders        10  New Customer  6.367797   Low Spender  

[5 rows x 24 columns]

## Frequency Flag

#### Question 7: Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
* If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
* If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
* If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [25]:
# creating new column for median order duration
df['median_order_duration'] = df.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [26]:
# creating a flag for customers whose median order duration is greater than 20 as 'Non-Frequent Customer'
df.loc[df['median_order_duration'] > 20, 'frequency_flag'] = 'Non-Frequent Customer'

In [27]:
# creating a flag for customers whose median order duration is greater than 10 but less than/equal to 20 as 'Regular Customer'
df.loc[(df['median_order_duration'] > 10) & (df['median_order_duration'] <= 20), 'frequency_flag'] = 'Regular Customer'

In [28]:
# creating a flag for customers whose median order duration is less than or equal to 10 as 'Frequent Customer'
df.loc[df['median_order_duration'] <= 10, 'frequency_flag'] = "Frequent Customer"

In [29]:
df['frequency_flag'].value_counts(dropna=False)

Frequent Customer        21560324
Regular Customer          7209005
Non-Frequent Customer     3636707
NaN                             5
Name: frequency_flag, dtype: int64

#### Customers who have a median order duration of 10 or less are the majority of customers.  

In [30]:
#removing maximum number of roles for display
pd.options.display.max_rows = None

In [31]:
df.head(100)

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0    2539329        1    prior             1          2                  8   
1    2398795        1    prior             2          3                  7   
2     473747        1    prior             3          3                 12   
3    2254736        1    prior             4          4                  7   
4     431534        1    prior             5          4                 15   
5    3367565        1    prior             6          2                  7   
6     550135        1    prior             7          1                  9   
7    3108588        1    prior             8          1                 14   
8    2295261        1    prior             9          1                 16   
9    2550362        1    prior            10          4                  8   
10   2968173       15    prior            15          1                  9   
11   1870022       15    prior            17          2                 16   
12   1911383       15    prior            18          2                 11   
13   2715276       15    prior            21          1                  9   
14    487368       15    prior            22          1                 10   
15   2293453       19    prior             2          5                 14   
16   1973799       19    prior             5          6                 12   
17    532817       19    prior             7          4                 17   
18   1573906       21    prior            10          3                 10   
19   1593000       31    prior            10          3                  8   
20   2231262       31    prior            17          3                 11   
21   2580647       43    prior             6          4                 16   
22   2187180       43    prior             9          4                 12   
23   2497897       52    prior             1          1                  9   
24   1318871       52    prior             2          1                 10   
25   1261384       52    prior             3          1                 10   
26    580568       52    prior             5          2                 10   
27   2428073       52    prior             7          2                  7   
28    342306       52    prior             9          1                 11   
29   1520353       52    prior            10          2                 10   
30   2542086       52    prior            11          2                  9   
31    944694       52    prior            12          1                 16   
32   1268191       52    prior            15          2                 11   
33   1122089       52    prior            17          2                 10   
34   1498922       52    prior            18          2                 10   
35    180919       52    prior            19          3                  8   
36   2695875       52    prior            21          5                 10   
37   3205960       67    prior             1          3                 11   
38    970922       67    prior             2          1                 15   
39    596585       67    prior             3          1                 13   
40   2037087       67    prior             4          2                  9   
41   2474111       67    prior             5          3                 14   
42   1568481       67    prior             6          3                 12   
43   2048527       67    prior             8          2                 11   
44   2389031       67    prior             9          2                 14   
45   2378562       67    prior            10          1                 10   
46   1515698       67    prior            11          1                 11   
47   1928050       67    prior            12          2                 12   
48   1949274       67    prior            13          1                 10   
49    551869       67    prior            14          1                  9   
50    768572       67    p

#### Extra Step: Checking missing values to see why there are blanks

In [32]:
df_blank = df[df['frequency_flag'].isnull() == True]

In [33]:
df_blank

order_id  user_id eval_set  order_number  order_dow  \
13645692    895835   159838    prior             1          0   
17251990    895835   159838    prior             1          0   
17622767    895835   159838    prior             1          0   
24138593    895835   159838    prior             1          0   
25881080    895835   159838    prior             1          0   

          order_hour_of_day  days_since_prior_order  product_id  \
13645692                 17                     NaN       10749   
17251990                 17                     NaN       33401   
17622767                 17                     NaN       23695   
24138593                 17                     NaN       21334   
25881080                 17                     NaN       22198   

          add_to_cart_order  reordered  ...    price_range_loc  Busiest day  \
13645692                  3          0  ...  Mid-range product  Busiest day   
17251990                  6          0  ...  Mid-range product  Busiest day   
17622767                  2          0  ...  Low-range product  Busiest day   
24138593                  5          0  ...  Mid-range product  Busiest day   
25881080                  1          0  ...  Low-range product  Busiest day   

          Busiest days     Most Orders  max_order  loyalty_flag avg_price  \
13645692  Busiest days  Average Orders          1  New Customer      7.42   
17251990  Busiest days  Average Orders          1  New Customer      7.42   
17622767  Busiest days  Average Orders          1  New Customer      7.42   
24138593  Busiest days  Average Orders          1  New Customer      7.42   
25881080  Busiest days  Average Orders          1  New Customer      7.42   

         spending_flag median_order_duration frequency_flag  
13645692   Low Spender                   NaN            NaN  
17251990   Low Spender                   NaN            NaN  
17622767   Low Spender                   NaN            NaN  
24138593   Low Spender                   NaN            NaN  
25881080   Low Spender                   NaN            NaN  

[5 rows x 26 columns]

#### There are 5 blank values on the frequency check.  These 5 values correlate to first time shoppers only.  As they only have 1 order and do not have a days_since_prior order listed.  

In [34]:
# Question 9: Export dataframe as a pickle file and store it correctly in your "prepared data" folder
df.to_pickle(r'C:\Users\littl\Career Foundry\06.2022 Instacart Basket Analysis\02 - Data\Prepared Data\4.8_orders_products_merged.pkl')

#### Removing outliers

In [35]:
df['prices'].max()

99999.0

In [36]:
df.loc[df['prices']>100, 'prices'] = np.nan

In [37]:
df['prices'].max()

25.0

#### Exporting updated data without outliers

In [38]:
df.to_pickle(r'C:\Users\littl\Career Foundry\06.2022 Instacart Basket Analysis\02 - Data\Prepared Data\4.8_orders_products_merged.pkl')

In [39]:
df.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  ...  \
0                     NaN         196                  1          0  ...   
1                    15.0         196                  1          1  ...   
2                    21.0         196                  1          1  ...   
3                    29.0         196                  1          1  ...   
4                    28.0         196                  1          1  ...   

     price_range_loc     Busiest day  Busiest days     Most Orders  max_order  \
0  Mid-range product  Regularly busy  Regular days  Average Orders         10   
1  Mid-range product  Regularly busy  Slowest days  Average Orders         10   
2  Mid-range product  Regularly busy  Slowest days     Most Orders         10   
3  Mid-range product      Least busy  Slowest days  Average Orders         10   
4  Mid-range product      Least busy  Slowest days     Most Orders         10   

   loyalty_flag avg_price spending_flag median_order_duration  \
0  New Customer  6.367797   Low Spender                  20.5   
1  New Customer  6.367797   Low Spender                  20.5   
2  New Customer  6.367797   Low Spender                  20.5   
3  New Customer  6.367797   Low Spender                  20.5   
4  New Customer  6.367797   Low Spender                  20.5   

          frequency_flag  
0  Non-Frequent Customer  
1  Non-Frequent Customer  
2  Non-Frequent Customer  
3  Non-Frequent Customer  
4  Non-Frequent Customer  

[5 rows x 26 columns]